Zend/zend_API.c

```diff
@@ -3723,58 +3724,151 @@ ZEND_API int zend_declare_property_stringl(zend_class_entry *ce, const char *nam
 }
 /* }}} */
 
-ZEND_API int zend_declare_class_constant(zend_class_entry *ce, const char *name, size_t name_length, zval *value) /* {{{ */
+ZEND_API int zend_declare_class_constant_ex(zend_class_entry *ce, zend_string *name, zval *value, int access_type, zend_string *doc_comment) /* {{{ */
 {
+	zend_class_constant_info *const_info;
+
+	if (zend_string_equals_literal_ci(name, "class")) {
+		zend_error(E_CORE_ERROR,
+				"A class constant must not be called 'class'; it is reserved for class name fetching");
+	}
+
+	const_info = zend_hash_find_ptr(&ce->constants_info, name);
+	if (const_info == NULL) {
+		if (ce->type == ZEND_INTERNAL_CLASS) {
+			const_info = pemalloc(sizeof(zend_class_constant_info), 1);
+		} else {
+			const_info = zend_arena_alloc(&CG(arena), sizeof(zend_class_constant_info));
+		}
+
+		const_info->offset = ce->constants_count++;
+		const_info->name = zend_string_copy(name);
+		const_info->name = zend_new_interned_string(const_info->name);
+		const_info->flags = access_type;
+		const_info->doc_comment = doc_comment;
+		const_info->ce = ce;
+
+		ce->constants_table = perealloc(ce->constants_table, sizeof(zval) * ce->constants_count, ce->type == ZEND_INTERNAL_CLASS);
+		zend_hash_add_ptr(&ce->constants_info, name, const_info);
+	} else {
+		i_zval_ptr_dtor(&ce->constants_table[const_info->offset] ZEND_FILE_LINE_CC);
```

Comment:
The old behavior was to just update the value of the constant if it already existed, and some extensions in tree now depend on it (Date being one of them) 

I thought better to keep the existing behavior, because maybe extensions outside php-src depend on this?